### ISSUE TITLE
Model.find returns items with incorrect order

### ISSUE THREAD

**Issue by trongnd** [strongloop/loopback-datasource-juggler#800]
I'm using loopback for mongodb and facing a problem when trying **Model.find** with **order**:

```javascript
User.find({
    order: 'createdAt DESC',
    limit: 10,
    include: ['publisher']
});
```
The problem here is the order of returned items is incorrect. I have inspect some code I find out the problem caused by this code:

```javascript
 // master/lib/dao.js#L1455
async.each(data, function(item, callback) {
....
// master/lib/dao.js#L1502
results.push(obj);
```
Because **async.each** run in parallel so **results.push(obj)** is not guaranteed to be run in correct order. That why I get back a list of items in incorrect order.

I'm currently using with the code below to temporarily fix the problem in own project:
```javascript
 // master/lib/dao.js#L1455
async.forEachOf(data, function(item, index, callback) {
....
// master/lib/dao.js#L1502
results[index] = obj;
```
It's working and also keeps the advantage of run in  parallel of async.each.

Own project is going to release soon so we really hope this problem will be fixed asap.
Thank you so much!

**Comment by maximeblondeau** [comment]
Same issue for me, I used the mongo connector to solve this.

**Comment by bajtos** [comment]
The pull request #864 contains a unit-test exposing the problem:

```js
    it('should query sorted desc by order integer field even though there' +
       'is an async model loaded hook', function(done) {
      User.observe('loaded', function(ctx, next) {
        var randomTimeoutTrigger = Math.floor(Math.random() * 100);

        setTimeout(function() {
          process.nextTick(next);
        }, randomTimeoutTrigger);
      });

      User.find({ order: 'order DESC' }, function(err, users) {
        should.exists(users);
        should.not.exists(err);
        users.pop().order.should.equal(1);
        users.pop().order.should.equal(2);
        users.shift().order.should.equal(6);
        done();
      });
    });
```

I am labeling this issue as a major bug and moving to Top of Backlog to prioritise higher. I think the fix should be reasonably easy, hence `beginner-friendly`.

**Comment by bajtos** [comment]
Fixed in `loopback-datasource-juggler@2.47.0`
